lib/ipmi_user.c

```diff
@@ -626,42 +648,41 @@ ipmi_user_mod(struct ipmi_intf *intf, int argc, char **argv)
 	return 0;
 }
 
-#define USER_PW_IPMI15_LEN 16 /* IPMI 1.5 only allowed for 16 bytes */
-#define USER_PW_IPMI20_LEN 20 /* IPMI 2.0 allows for 20 bytes */
-#define USER_PW_MAX_LEN USER_PW_IPMI20_LEN
-
 int
 ipmi_user_password(struct ipmi_intf *intf, int argc, char **argv)
 {
-	char *password = NULL;
+	char *password;
+	char passbuf[USER_PW_MAX_LEN + 1];
+	char tmp[USER_PW_MAX_LEN + 1];
 	int ccode = 0;
 	uint8_t password_type = USER_PW_IPMI15_LEN;
 	size_t password_len;
 	uint8_t user_id = 0;
+	size_t tmplen;
+
 	if (is_ipmi_user_id(argv[2], &user_id)) {
 		return (-1);
 	}
 
 	if (argc == 3) {
 		/* We need to prompt for a password */
-		char *tmp;
-		size_t tmplen;
-		password = ask_password(user_id);
-		if (!password) {
-			lprintf(LOG_ERR, "ipmitool: malloc failure");
+		memset(passbuf, 0, sizeof(passbuf));
+		memset(tmp, 0, sizeof(tmp));
+
+		if (ask_password(user_id, passbuf, sizeof(passbuf))) {
 			return (-1);
 		}
-		tmp = ask_password(user_id);
-		tmplen = strnlen(tmp, USER_PW_MAX_LEN + 1);
-		if (!tmp) {
-			lprintf(LOG_ERR, "ipmitool: malloc failure");
+		if (ask_password(user_id, tmp, sizeof(tmp))) {
 			return (-1);
 		}
-		if (strncmp(password, tmp, tmplen)) {
+		tmplen = strnlen(tmp, USER_PW_MAX_LEN + 1);
+		if (strncmp(passbuf, tmp, tmplen)) {
 			lprintf(LOG_ERR, "Passwords do not match or are "
 			                 "longer than %d", USER_PW_MAX_LEN);
 			return (-1);
 		}
+		/* XXX memcpy out passwords asap. */
```

Comment:
I don't exactly understand this comment here. Who and when do you want to do this?

Reply:
It's more or less a warning that a without memset/bzeroing the password buffers immediately after use, the user password will be resident in memory. It's likely a low risk attack vector, since ipmitool exits shortly after invocation in this case, however on an idle system, if the virtual memory pages are not reused rather quickly, in theory the passwords could be read by someone with know how.

Reply:
Yes, that is clear, I'm just wondering who is the target of this comment. Who exactly needs to do this, when they are going to do this (never, I figure), and if it needs to be done, then why not do it right away?